src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.cs

```diff
@@ -1,53 +1,85 @@
 {
-    internal abstract class AbstractDiagnosticsTaggerProvider<TTag> :
-        AbstractAsynchronousTaggerProvider<AbstractAggregatedDiagnosticsTagSource<TTag>, TTag>, ITaggerProvider
-        where TTag : ITag
+    internal abstract class AbstractDiagnosticsTaggerProvider<TTag> : AsynchronousTaggerProvider<TTag> where TTag : ITag
     {
-        protected readonly DiagnosticService DiagnosticService;
+        private readonly IDiagnosticService _diagnosticService;
 
-        public AbstractDiagnosticsTaggerProvider(
+        protected AbstractDiagnosticsTaggerProvider(
             IDiagnosticService diagnosticService,
             IForegroundNotificationService notificationService,
             IAsynchronousOperationListener listener)
             : base(listener, notificationService)
         {
-            this.DiagnosticService = diagnosticService as DiagnosticService;
+            _diagnosticService = diagnosticService;
         }
 
-        public ITagger<T> CreateTagger<T>(ITextBuffer subjectBuffer) where T : ITag
-        {
-            if (subjectBuffer == null)
-            {
-                throw new ArgumentNullException(nameof(subjectBuffer));
-            }
+        protected abstract bool IsEnabled { get; }
+        protected abstract bool IncludeDiagnostic(DiagnosticData data);
+        protected abstract ITagSpan<TTag> CreateTagSpan(SnapshotSpan span, DiagnosticData data);
 
-            return this.GetOrCreateTagger<T>(null, subjectBuffer);
-        }
-
-        internal sealed override bool TryRetrieveTagSource(ITextView textViewOpt, ITextBuffer subjectBuffer, out AbstractAggregatedDiagnosticsTagSource<TTag> tagSource)
+        public override sealed ITaggerEventSource CreateEventSource(ITextView textViewOpt, ITextBuffer subjectBuffer)
         {
-            return subjectBuffer.Properties.TryGetProperty(UniqueKey, out tagSource);
+            return TaggerEventSources.Compose(
+                TaggerEventSources.OnWorkspaceRegistrationChanged(subjectBuffer, TaggerDelay.Medium),
+                TaggerEventSources.OnDiagnosticsChanged(subjectBuffer, _diagnosticService, TaggerDelay.Medium));
         }
 
-        protected sealed override void StoreTagSource(ITextView textViewOpt, ITextBuffer subjectBuffer, AbstractAggregatedDiagnosticsTagSource<TTag> tagSource)
+        public override sealed Task ProduceTagsAsync(TaggerContext<TTag> context, DocumentSnapshotSpan spanToTag, int? caretPosition)
         {
-            subjectBuffer.Properties.AddProperty(UniqueKey, tagSource);
+            ProduceTags(context, spanToTag);
+            return SpecializedTasks.EmptyTask;
         }
 
-        protected sealed override void RemoveTagSource(ITextView textViewOpt, ITextBuffer subjectBuffer)
+        private void ProduceTags(TaggerContext<TTag> context, DocumentSnapshotSpan spanToTag)
         {
-            subjectBuffer.Properties.RemoveProperty(UniqueKey);
+            if (!IsEnabled)
+            {
```

Comment:
@CyrusNajmabadi ya, which will report diagnostics for every diagnostic providers for the document, rather than each one of them separately.